source/custom/visa_service.custom.cpp

```diff
@@ -467,6 +469,56 @@ ::grpc::Status VisaService::ParseRsrc(::grpc::ServerContext* context, const Pars
   }
 }
 
+//---------------------------------------------------------------------
+//---------------------------------------------------------------------
+::grpc::Status VisaService::PxiReserveTriggers(::grpc::ServerContext* context, const PxiReserveTriggersRequest* request, PxiReserveTriggersResponse* response)
+{
+  if (context->IsCancelled()) {
+    return ::grpc::Status::CANCELLED;
+  }
+  try {
+    auto vi_grpc_session = request->vi();
+    ViSession vi = session_repository_->access_session(vi_grpc_session.name());
+    auto cnt_determine_from_sizes = std::array<int, 2>
+    {
+      request->trig_buses_size(),
+      request->trig_lines_size()
+    };
+    const auto cnt_size_calculation = calculate_linked_array_size(cnt_determine_from_sizes, false);
+
+    if (cnt_size_calculation.match_state == MatchState::MISMATCH) {
+      return ::grpc::Status(::grpc::INVALID_ARGUMENT, "The sizes of linked repeated fields [trig_buses, trig_lines] do not match");
+    }
+    auto cnt = cnt_size_calculation.size;
+
+    auto trig_buses_request = request->trig_buses();
+    std::vector<ViInt16> trig_buses;
+    std::transform(
+      trig_buses_request.begin(),
+      trig_buses_request.end(),
+      std::back_inserter(trig_buses),
+      [](auto x) { return (ViInt16)x; }); 
+    auto trig_lines_request = request->trig_lines();
+    std::vector<ViInt16> trig_lines;
+    std::transform(
+      trig_lines_request.begin(),
+      trig_lines_request.end(),
+      std::back_inserter(trig_lines),
+      [](auto x) { return (ViInt16)x; }); 
+    ViInt16 failure_index {};
+    auto status = library_->PxiReserveTriggers(vi, cnt, trig_buses.data(), trig_lines.data(), &failure_index);
+    if (!status_ok(status)) {
+      context->AddTrailingMetadata("ni-failing-index", std::to_string(failure_index));
```

Comment:
The doc you called out in your email said:

> if this function returns an error code directly related to reserving triggers. [...]. For any other status code returned, the value of this output parameter is undefined and should not be used.

So it seems this code would have Undefined Behavior for any non-trigger-related errors. Is it practical to trap those specific errors? Or maybe by code inspection decide that it's not really UB? Also, if those non-trigger-error cases return `-1`, would you prefer to have `ni-failing-index` be unset in that case to be more consistent with success?

Related note: should `failure_index` be initialized to `-1`?

Reply:
It's not practical to look at specific error codes because vendor-specific error codes can always extend it. This is called so rarely that it's not a performance concern. Returning it via the metadata in all cases seems reasonable to me.